include/SampleBuffer.h

```diff
@@ -51,17 +52,52 @@ class LMMS_EXPORT SampleBuffer
 	using reverse_iterator = std::vector<sampleFrame>::reverse_iterator;
 	using const_reverse_iterator = std::vector<sampleFrame>::const_reverse_iterator;
 
-	SampleBuffer() = default;
-	explicit SampleBuffer(const QString& audioFile);
-	SampleBuffer(const QString& base64, int sampleRate);
-	SampleBuffer(std::vector<sampleFrame> data, int sampleRate);
-	SampleBuffer(
-		const sampleFrame* data, int numFrames, int sampleRate = Engine::audioEngine()->processingSampleRate());
+	enum class Source
+	{
+		Unknown,
+		AudioFile,
+		Base64
+	};
+
+	//! passkey idiom
+	class Access
+	{
+	public:
+		friend class SampleBuffer;
+		friend class SampleLoader;
+		Access(Access&&) = default;
+	private:
+		Access() {}
+		Access(const Access&) = default;
+	};
```

Comment:
The `create()` methods, passkey idiom, and `std::enable_shared_from_this` all work to guarantee that all `SampleBuffer` objects are referred to by shared pointers and cannot be created any other way. I was originally thinking this would be important because my initial idea for a cache eviction implementation (before I settled on a custom deleter for `std::shared_ptr`) involved calling `weak_from_this()` or `shared_from_this()` in the `SampleBuffer`.

But now maybe there isn't much point to this. And maybe we'd want to be able to, say, create a plain `SampleBuffer` on the stack. So I'll look into removing it.